gateway-server/server.py

```diff
@@ -0,0 +1,91 @@
+from flask import Flask, request, jsonify, render_template
+import requests
+# TODO: from kazoo.client import KazooClient
+from pymongo import MongoClient
+import random
+
+app = Flask(__name__)
+
+# TODO: Add Zookeeper functionality
+
+mongo_uri = '<hidden-for-pull-request>'
+client = MongoClient(mongo_uri)
+db = client['nabu']
+
+# List of IPFS node addresses
+IPFS_URL = [
+    'http://localhost:5001',
+    # TODO: Add more IPFS node addresses
+]
+
+# IPFS routes
+PUT_ROUTE = '/api/v0/block/put'
+GET_ROUTE = '/api/v0/block/get'
+
+# Web UI
+@app.route('/')
+def index():
+    return render_template('index.html')
+
+# Forward GET request to IPFS peer to retrieve all content
+@app.route('/ipfs', methods=['GET'])
+def get_ipfs_content():
+    # Retrieve CIDs from database
+    cid_records = db['cid'].find()
```

Comment:
So, you are hitting the database on every GET request. By doing so, you may get away with ZooKeeper completely and have multiple stateless replicas behind a simple load balancer.

Reply:
Do you think it'll be better to keep a local copy of the CID list and only read when the leader changes?